### ISSUE TITLE
Separate query builder into its own library

### ISSUE THREAD

**Issue by dantleech** [doctrine/phpcr-odm#627]
The [QueryBuilder](https://github.com/doctrine/phpcr-odm/tree/master/lib/Doctrine/ODM/PHPCR/Query/Builder) could be used for plain PHPCR (e.g. replacing the PHPCR query builder) in addition to other PHPCR based document managers (I have started a [light-weight DocumentManager](https://github.com/sulu-io/sulu/pull/972) for Sulu).

I propose to make it a `phpcr` library with extension points (e.g. introduce a factory to create the query nodes). 

In principle it could also be used to create standard RDBMS queries, but perhaps this would be pushing the boat out too far.

**Comment by dbu** [comment]
sounds like a good idea if there is potential to reduce duplication. is there any chance we could make the querybuilder shareable with doctrine orm and the other doctrine flavours? that would have a huge potential of reducing duplication. of course, storage specific options (advanced SQL features, PHPCR specific child-parent relations, ...) would need to be added in extension points, but the basic infrastructure and from, where, and so on could be shared i think. @Ocramius wdyt?

**Comment by dantleech** [comment]
The only "issue" would be that this query builder uses is completely fluent:

````
$qb->select()->column('foo')->column('bar')->from()->document('Foobar')->end()->where()->eq()->field('bar')->litreal('foo');
````

Which is a big departure from the standard Doctrine QueryBuilder, but was necessary to avoid a nasty factory interface ( `$qb->from($factory->sourceInnerJoin($factory->sourceDocument('asdf'), $factory->conditionalJoin(...`).

**Comment by dbu** [comment]
maybe the orm could provide both in parallel and deprecate their factory based one?

**Comment by lsmith77** [comment]
+1 from me to split it out

**Comment by dbu** [comment]
@dantleech was #638 a first step in this direction?

**Comment by dantleech** [comment]
A crude step maybe. Sulu builds on and extends the QB but it is not that elegant. Of course would be good for Sulu if it were a separate lib but it might need to be rearchitectured if it were to be a truly generic QB.



On 4 July 2015 12:43:07 CEST, David Buchmann <notifications@github.com> wrote:
>@dantleech was #638 a first step in this direction?
>
>---
>Reply to this email directly or view it on GitHub:
>https://github.com/doctrine/phpcr-odm/issues/627#issuecomment-118499581

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by dbu** [comment]
should we move this issue to doctrine commons instead of here?

**Comment by lsmith77** [comment]
ping

**Comment by dbu** [comment]
closing due to inactivity
